File: src/main/java/at/ac/tuwien/complang/recursive/NQueensSolverRecursive3.java
package at.ac.tuwien.complang.recursive;

/**
 * Based on OriginalNQueensSolver.
 * 
 * Removed field x since there is no real use for it
 * Removed maxN and used n instead
 * Common subexpressions eliminated
 *
 */
public class NQueensSolverRecursive3 {

    private final int n;
    
    // 1..n
    // a[x] ist true, wenn man in Spalte x eine Dame setzen kann
    private final boolean[] a;
    
    // 2..2*n
    // b[x+y] ist true, wenn man in der Diagonale, 
    // die die Felder (x+k, y-k) (fuer alle k) enthaelt, eine Dame setzen kann
    private final boolean[] b;
    
    // 1..2*n-1
    // c[x-y+n] ist true, wenn man in der Diagonale,
    // die die Felder (x+k, y+k) (fuer alle k) enthaelt, eine Dame setzen kann
    private final boolean[] c;

    private long count = 0;
    
    public NQueensSolverRecursive3(int n) {
        this.n = n;
        this.a = new boolean[n + 1];
        this.b = new boolean[2 * n + 1];
        this.c = new boolean[2 * n];
        int i;

        for (i = 1; i <= n; i++) {
            a[i] = true;
        }
        for (i = 2; i <= 2 * n; i++) {
            b[i] = true;
        }
        for (i = 1; i <= 2 * n - 1; i++) {
            c[i] = true;
        }
    }

    private void try1(int i) {
        int j = 0;
        int bIndex = i;
        int cIndex = i + n;
        do {
            j++;
            bIndex++;
            cIndex--;
            if (a[j] && b[bIndex] && c[cIndex]) {
                if (i < n) {
                    a[j] = false;
                    b[bIndex] = false;
                    c[cIndex] = false;
                    try1(i + 1);
                    a[j] = true;
                    b[bIndex] = true;
                    c[cIndex] = true;
                } else {
                    count++;
                }
            }
        } while (j != n);
    }

    public long solve() {
        try1(1);
        return count;
    }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        NQueensSolverRecursive3 o = new NQueensSolverRecursive3(n);
        System.out.println(o.solve());
    }
}

File: src/main/c/NQueensSolverRecursive3.c
/******************************************************************************
 * Based on OriginalNQueensSolver.
 *
 * Removed field x since there is no real use for it
 * Removed maxN and used n instead (use malloc to reserve int arrays on heap)
 * Common subexpressions eliminated
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>

/* this is the actual board size to be used */
int n;
/* number of found solutions to the n-queens problem */
long count = 0L;
/* a[x] is true, if a queen can be placed in column x  (x: 1..n) */
int *a;
/* b[y+x] is true, if a queen can be placed in the diagonal which is covered
 by the fields (x+k, y-k) for all k.
 "/" antidiagonal line) */
int *b;
/* c[y-x+n] is true, if a queen can be placed in the diagonal which is covered
 by the fields (x+k, y+k) for all k.
 "\" diagonal line */
int *c;

/*
 * Search all solutions that start with a queen in the row i.
 */
void try1(int i) {
	int j = 0;
	// Common subexpressions
    int bIndex = i;
    int cIndex = i + n;
	do {
        j++;
        bIndex++;
        cIndex--;
		if (a[j] && b[bIndex] && c[cIndex]) {
			a[j] = 0;
			b[bIndex] = 0;
			c[cIndex] = 0;
			if (i < n) {
				try1(i + 1);
			} else {
				count++;
			}
			a[j] = 1;
			b[bIndex] = 1;
			c[cIndex] = 1;
		}
	} while (j != n);
}

int main(int argc, char *argv[]) {
	if (argc <= 1 || (n = atoi(argv[1])) <= 0) {
		n = 8;
	}
	// make arrays the required size
	a = (int *) malloc((n + 1) * sizeof(int));
	b = (int *) malloc((2 * n + 1) * sizeof(int));
	c = (int *) malloc((2 * n) * sizeof(int));
	// initialize all arrays, set all true
	int i = 1;
	for (i = 1; i <= n; i++) {
		a[i] = 1; // '|' column storage
	}
	for (i = 2; i <= (2 * n); i++) {
		b[i] = 1; // '/' diagonal storage
	}
	for (i = 1; i <= ((2 * n) - 1); i++) {
		c[i] = 1; // '\' diagonale storage
	}
	try1(1);
	printf("%ld\n", count);
	return 0;
}

